faculty_distributed/__init__.py

```diff
@@ -0,0 +1,149 @@
+import cloudpickle
+import distributed
+import os
+import time
+import shutil
+import tempfile
+
+from faculty import client
+
+
+class FacultyJobExecutor:
+    """
+    Run generic python function in parallel on Faculty Jobs
+
+    Parameters
+    ----------
+    project_id: str
+        ID of project on Faculty platform.
+    job_id: str
+        ID of job set up in Jobs tab of Faculty platform.
+    clean: bool, default=True
+        If True, temporary directory used to pass files to and from workers
+        will be removed once workers are finished. If False, temporary files
+        will be left on the workspace.
+    tmpdir_prefix: str, default=None
+        Root directory for temporary directory to be created in. If None,
+        temporary directory will be created in "/project/.faculty-distributed".
+    """
+
+    def __init__(self, project_id, job_id, clean=True, tmpdir_prefix=None):
+        self.project_id = project_id
+        self.job_id = job_id
+        self.job_client = client("job")
+
+        self.clean = clean
+        self.tmpdir_prefix = tmpdir_prefix
+
+    def _collect_output(self, args_sequence):
+        """
+        Collect output of jobs
+
+        Parameters
+        ----------
+        args_sequence: list
+            list of lists of arguments that were iterated over in parallel
+
+        Returns
+        -------
+        out: list
+            list of outputs of function from each job
+        """
+        out = []
+        for i in range(len(args_sequence)):
+            with open(
+                os.path.join(self.tmpdir, "output/out_{}.pkl".format(i)), "rb"
+            ) as f:
+                out.append(cloudpickle.load(f))
+
+        return out
+
+    def _make_dirs(self):
+        """Make temporary directories"""
+        if self.tmpdir_prefix is None:
+            self.tmpdir = tempfile.mkdtemp(
+                prefix="/project/.faculty-distributed"
+            )
+        else:
+            self.tmpdir = tempfile.mkdtemp(prefix=self.tmpdir_prefix)
+
+        for jobs_dir in [
+            os.path.join(self.tmpdir, "output"),
+            os.path.join(self.tmpdir, "func"),
+        ]:
+            os.makedirs(jobs_dir)
+
+    def map(self, func, args_sequence):
```

Comment:
It's fairly customary to group public methods either at the top (below the constructor) or at the bottom.